### PR TITLE
#1 move jailing code to core

### PR DESCRIPTION
**Pull request by IljaN**
## Description
- Read-only jailing of guest users is now provided by core
- Read-only groups can be configured in oc_appconfig under 'core' => 'read_only_groups' (json array)

## Related Issue
#https://github.com/owncloud/guests/issues/1

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: settings/Controller/UsersController.php [hunk 1]
```diff
@@ -244,19 +246,42 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
...
+						$limit,
+						$offset
+					)
+				);
 			} else {
 				$batch = $this->userManager->search($pattern, $limit, $offset);
+				if (!empty($excludeGroups) && !empty($batch)) {
+					$excludeGroups = explode(';', $excludeGroups);
+					$excludedUsers = [];
+					foreach ($excludeGroups as $excludedGroup) {
```
Comment (PVince81) [inline settings/Controller/UsersController.php]: so here you're post-filtering.

From what I see `displayNamesInGroup` was the function that did the filtering so maybe we should extend that one to include a "excludeGroups parameter" so we can optimize later on DB level.

Or we just leave the post-filtering if not too expensive...

Comment (IljaN) [inline settings/Controller/UsersController.php]: @PVince81 Now that i see it, should i remove this code again? It is not need atm anyway, and displayNamesInGroup is probably the better place for it.

Comment (PVince81) [inline settings/Controller/UsersController.php]: yeah, keep it for another PR maybe

File: lib/private/Files/Storage/Wrapper/DirMask.php [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OC\Files\Storage\Wrapper;
+
+use OC\Files\Storage\Wrapper\PermissionsMask;
+
+class DirMask extends PermissionsMask {
```
Comment (PVince81) [inline lib/private/Files/Storage/Wrapper/DirMask.php]: add PHPDoc, why do we need DirMask and how is it different from PermissionsMask ?

File: lib/private/Files/Storage/Wrapper/DirMask.php [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	public function isDeletable($path) {
+		if ($this->checkPath($path)) {
+			return parent::isDeletable($path);
+		} else {
+			return $this->storage->isDeletable($path);
+		}
+	}
+
+	public function isSharable($path) {
+		if ($this->checkPath($path)) {
```
Comment (PVince81) [inline lib/private/Files/Storage/Wrapper/DirMask.php]: If we do decide to prevent sharing globally for guests, this is the place where we could always return `false`.

Let's leave this as is for now.

File: lib/private/Files/Storage/Wrapper/DirMask.php [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	public function getPermissions($path) {
+		if ($this->checkPath($path)) {
+			return parent::getPermissions($path);
+		} else {
+			return $this->storage->getPermissions($path);
+		}
+	}
+
+	public function rename($path1, $path2) {
+		if (!$this->isReadable($path1)) {
```
Comment (PVince81) [inline lib/private/Files/Storage/Wrapper/DirMask.php]: does this need special handling for when $path1 is a part file and $path2 isn't ?

File: lib/private/Files/Storage/Wrapper/DirMask.php [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	 * @param string $path
+	 * @param \OC\Files\Storage\Storage (optional) the storage to pass to the cache
+	 * @return \OC\Files\Cache\Cache
+	 */
+	public function getCache($path = '', $storage = null) {
+		if (!$storage) {
+			$storage = $this;
+		}
+		if ($this->checkPath($path)) {
+			return parent::getCache($path, $storage);
```
Comment (PVince81) [inline lib/private/Files/Storage/Wrapper/DirMask.php]: hmm, is this method override really needed ?

File: lib/private/User/Session.php [hunk 6]
```diff
@@ -395,8 +396,15 @@ protected function prepareUserLogin($firstTimeLogin) {
...
 			$userFolder = \OC::$server->getUserFolder($user);
 
-			// copy skeleton
-			\OC_Util::copySkeleton($user, $userFolder);
+			try {
+				// copy skeleton
+				\OC_Util::copySkeleton($user, $userFolder);
+			} catch (NotPermittedException $ex) {
+				// possible if files directory is in an readonly jail
+				\OC::$server->getLogger()->info(
```
Comment (PVince81) [inline lib/private/User/Session.php]: would "debug" be better here or "warning" ?

File: lib/private/legacy/util.php [hunk 7]
```diff
@@ -224,6 +224,51 @@ public static function setupFS($user = '') {
...
+		);
+
+		$isGuest = \OC::$server->getConfig()->getUserValue(
+			$user,
+			'owncloud',
+			'isGuest',
+			false
+		);
+
+		if ($isGuest === '1' || !empty($readOnlyGroupMemberships)) {
```
Comment (PVince81) [inline lib/private/legacy/util.php]: I suggest to only retrieve `$readOnlyGroupMemberships` and the related logic if `$isGuest` is true. Because in most cases the related logic does not need to run at all (and potentially slow down the code a few microseconds...)

**Review by PVince81 — COMMENTED** [review]
Code looks good otherwise.

File: lib/private/legacy/util.php (hunk 7 above)
Comment (IljaN) [inline lib/private/legacy/util.php]: You probably mean if $isGuest is false. if $isGuest is true we know for sure that we need the jail thus don`t need to concern ourselves with readonly groups.

File: lib/private/Files/Storage/Wrapper/DirMask.php (hunk 4 above)
Comment (IljaN) [inline lib/private/Files/Storage/Wrapper/DirMask.php]: Why would it?

Comment (PVince81) [inline lib/private/Files/Storage/Wrapper/DirMask.php]: I saw later on that you added the special handling in the subclass, ok

File: lib/private/legacy/util.php (hunk 7 above)
Comment (PVince81) [inline lib/private/legacy/util.php]: ah yes... I forgot that these are two separate mechanisms

**Comment by PVince81** [comment]
Ready for rereview or anything remains to do ?

**Comment by IljaN** [comment]
Yep ready

**Comment by PVince81** [comment]
Can we have some unit tests for the new classes ?

Integration tests would be nice too but can be done separately. Can you provide rough steps how to set this up for integration tests ?

**Comment by PVince81** [comment]
(the main code looks good)

**Comment by IljaN** [comment]
There are two cases:
- User is guest by flag => create entry "isGuest=1" for the user in oc_preferences


- Create a group for example "guests" , add user to this group. Create a entry in oc_appconfig which looks like this: 
core	read_only_groups 	["guest_users"]

| appid | configkey | configvalue |
| ------------- | ------------- |-----------|
| core  | read_only_groups  | ["guests"] |

The outcome of both cases should be that the user is in a "Readonly Jail" and can`t upload files.
If a directory is shared with the user he can do everything in the folder according to the share permissions.

**Comment by PVince81** [comment]
Looks good, thanks. 👍

File: lib/private/legacy/util.php [hunk 8]
```diff
@@ -224,6 +224,57 @@ public static function setupFS($user = '') {
...
+			), true);
+
+			if (!is_array($readOnlyGroups)) {
+				$readOnlyGroups = [];
+			}
+
+
+			$userGroups = array_keys(
+				\OC::$server->getGroupManager()->getUserIdGroups($user)
+			);
```
Comment (mrow4a) [inline lib/private/legacy/util.php]: You are aware, that this introduced additional query - for each request going to oc, even most trivial? https://github.com/owncloud/core/issues/29380 @PVince81 @DeepDiver1975

Comment (mrow4a) [reply lib/private/legacy/util.php]: I mean, because isInGroup is calling exactly this function, it is not a problem (since it will be cached), but with central group table it might change (isInGroup should not fetch all groups of the users and then filter on app level)

Comment (mrow4a) [reply lib/private/legacy/util.php]: https://github.com/owncloud/core/pull/29384

Comment (PVince81) [reply lib/private/legacy/util.php]: >  that this introduced additional query

Are you saying this was added recently and is a regression ? Or is it part of the years old code ? Can you git blame and link to the PR that introduced it ?
